scrapbook/managers.py

```diff
@@ -0,0 +1,446 @@
+# -*- coding: utf-8 -*-
+"""
+managers.py
+
+Provides the encoders and stores for various data types to be persisted / recalled
+"""
+import six
+import json
+import base64
+import pandas as pd
+import collections
+
+from io import BytesIO
+from collections import OrderedDict
+
+# We lean on papermill's readers to connect to remote stores
+try:
+    from papermill.s3 import S3
+except ImportError:
+    # TODO: Handle missing depedency as papermill does
+    class S3(object): pass
+
+from .scraps import scrap_to_payload
+from .exceptions import ScrapbookMissingEncoder, ScrapbookMissingStore
+
+
+class DataManagerRegistry(collections.MutableMapping):
+    def __init__(self):
+        self._managers = OrderedDict()
+
+    def __getitem__(self, key):
+        return self._managers.__getitem__(key)
+
+    def __setitem__(self, key, value):
+        return self._managers.__setitem__(key, value)
+
+    def __delitem__(self, key):
+        return self._managers.__delitem__(key)
+
+    def __iter__(self):
+        return self._managers.__iter__()
+
+    def __len__(self):
+        return self._managers.__len__()
+
+    def register(self, encoder_name, store_name, manager):
+        """
+        Registers a new name to a particular encoder
+
+        Parameters
+        ----------
+        encoder_name: str (or None)
+            Name of the mime subtype parsed by the encoder. None implies any.
+        store_name: str (or None)
+            The specific storage type that this encoder supports. None implies any.
+        encoder: obj
+            The object which implements the required encoding functions.
+        """
+        # Add the registered name so we can reverse look-up these fields
+        manager.encoder_name = encoder_name
+        manager.store_name = store_name
+        self[(encoder_name, store_name)] = manager
+
+    def register_encoder(self, name, encoder):
+        self.register(name, None, encoder)
+
+    def register_store(self, name, store):
+        self.register(None, name, store)
+
+    def deregister(self, encoder_name, store_name):
+        """
+        Removes a particular encoder from the registry
+
+        Parameters
+        ----------
+        name: str
+            Name of the mime subtype parsed by the encoder.
+        store: str (optional)
+            The specific storage type that this encoder supports.
+        """
+        del self[(encoder_name, store_name)]
+
+    def reset(self):
+        """
+        Resets the registry to have no encoders.
+        """
+        self._managers = OrderedDict()
+
+    def find_manager(self, scrap):
+        if scrap.encoder is not None and scrap.store is not None:
+            return self.get((scrap.encoder, scrap.store))
+
+        for manager in reversed(list(self.values())):
+            # Only check for managers here that can both encode and store
+            if not (manager.encoder_name and manager.store_name):
```

Comment:
I found this function a little hard to read. I feel like it would be easier to read if this if-statement was inverted and the logic inside the try statement was placed inside it. Something like:

```python
if manager.encoder_name and manager.store_name:
    blah
```

Assuming I understand that the intent of this function is to run through the list of valid managers and retrieve the ones that work with the scrap passed as a parameter.

Reply:
Good point. I'll take a close look at this section next pass